### ISSUE TITLE
Training parameter is always default (True) in DCGAN lesson 

### ISSUE THREAD

**Issue by tupini07** [udacity/deep-learning#130]
Up until now we've seen that, when using `tf.layers.batch_normalization`, we set it's second parameter to True or False depending on whether we are training or not. However, I noticed that in the DCGAN class this parameter is always `True` since we are not passing any specific value when running the session (so it keeps the default value specified in the method definition). Is there any reason for this or is this an issue?

**Comment by clapollo** [comment]
I haven't looked at this notebook, but it doesn't seem to do batch normalization correctly. The way it's written now it **requires** training to be True because it never updates the population statistics. But I need to run through this today to see if that was intentional.

**Comment by mcleonard** [comment]
So I tried setting batch norm's `training=False` in the training process, but the generator wouldn't actually generate good images. From the documentation, it says setting `training=False` takes a moving average of the batches. This makes sense when you're just doing inference. But when you're training, the weights are changing between batches so the moving average gets messed up. I believe you want to keep `training=True` when generating images during training, but set `training=False` after training.

At least, this is what I found when developing the DCGAN model.

**Comment by clapollo** [comment]
First, you should be on vacation. :|

Second, I think the real problem is in the generator. It should set training to False when it's not training, like when we show the sample images. I think the discriminator in our examples is only ever used during training, so it doesn't matter that it's always set to True, but if you want to use the discriminator for something later, like a classifier, then you'd need to set it to False there, too. 

And just to clarify, setting training to True _calculates_ the moving averages, and setting it to False _uses_ the averages it calculated during training. You want the network to learn good population statistics during training so it can use them later.

I'll be making changes to the notebook tomorrow because there are still errors in how the batch normalization is implemented, even if it doesn't seem to change the results in _this_ notebook. Students are using this as a template for project 5, and the code as shown here does not work properly there.

**Comment by clapollo** [comment]
Ok, explained why it's ok to use training True for the discriminator in _this_ project, and changed so the generator sets training to False when sampling.
